design/resource-dependency.md

```diff
@@ -0,0 +1,348 @@
+The OAM runtime will build the dependency graph accordingly and only create dependent resources once the data outputs are ready.
+To achieve this, we propose:
+
+
+1. Add a field `dataOutputs` in Component configuration to specify a data output source:
+
+    ```yaml
+    dataOutputs:
+    - name: <dataoutput-name>
+      fieldPath: <object-field-path>
+    ```
+
+    For example, a my-rds database that outputs connection secret could be specified as:
+
+    ```yaml
+    kind: Component
+    metadata:
+      name: my-rds
+    spec:
+      workload:
+        ... # connection secret name will be written to 'status.connSecret'
+
+    ---
+    kind: ApplicationConfiguration
+    spec:
+      components:
+      - componentName: my-rds
+        dataOutputs:
+        - name: mysql-conn-secret
+          fieldPath: status.connSecret
+    ```
+
+2. Add a field `valueFrom.dataOutputName` in Component ParameterValue to specify the dependency on data outputs:
+
+    ```yaml
+    parameterValues:
+    - name: <parameter-name>
+      valueFrom:
+        dataOutputName: <dataoutput-object-name>
+    ```
+
+    For example, web app's dependency on RDS's database secret could be specified as:
+
+    ```yaml
+    kind: Component
+    metadata:
+      name: my-web
+    spec:
+      workload:
+        kind: MyWorkload
+        metadata:
+          name: my-web
+      parameters:
+      - name: database-secret # workload's fieldPaths needs to be filled
+        fieldPaths:
+        - spec.connSecret
+    ---
+
+    kind: ApplicationConfiguration
+    spec:
+      components:
+      - componentName: my-web
+        parameterValues:
+        - name: database-secret
+          valueFrom:
+            # Specify the dependency on a data output.
+            # The component's workload won't be created until the data output has returned value to fill workload's fieldPaths. 
+            dataOutputName: mysql-conn-secret
+
+      - componentName: my-rds
+        dataOutputs:
+        - name: mysql-conn-secret
+          fieldPath: status.connSecret
+    ```
+
+    In this case, the OAM runtime will automatically build a dependency from `my-web` to `my-rds`.
+
+    It will wait until `my-rds` object's field `status.connSecret` has value:
+
+    ```yaml
+    apiVersion: example.io/v1alpha1
+    kind: RDSInstance
+    metadata:
+      name: my-rds
+    status:
+      connSecret: mysql-conn # OAM runtime will wait for this field to have value
+    ```
+
+    Once the dependency has been satisfied, the OAM runtime will create `my-web` workload with
+    the corresponding field paths patched with values from data outputs.
+    
+    ```yaml
+    kind: MyWorkload
+    metadata:
+      name: my-web
+    spec:
+      connSecret: mysql-conn # patched by OAM runtime
+    ```
+
+3. Similary, add fields `dataOutputs` and `parameters` in Trait configuration to specify the dependency on data output objects.
```

Comment:
I get that since `trait` is supposedly maintained by `application operator` only, we don't need to have a `declare` phase and every field in a `trait` is parameterizable. I am not entirely convinced this is true though.